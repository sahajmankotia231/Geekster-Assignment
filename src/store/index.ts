import { createSlice, configureStore } from "@reduxjs/toolkit";

export type ChildType = {
  age: number | null;
};

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: ChildType[];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    initializeStore: (state, action): void => {
      if (Boolean(action.payload)) {
        const splitedRooms = action.payload?.split("|") || [];
        const rooms = splitedRooms?.map(
          (room): RoomType => {
            const guests = room.split(":");
            const adultsCounter = Number(guests[0]);
            const children =
              guests[1]?.split(",")?.reduce((arr, age): ChildType[] => {
                arr.push({ age: Number(age) });
                return arr;
              }, []) || [];

            return {
              ...initialRoomsState,
              adultsCounter,
              children,
            };
          }
        );

        state.rooms = rooms;
      }
    },
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  initializeStore,
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
