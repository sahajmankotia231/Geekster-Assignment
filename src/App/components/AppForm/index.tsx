import React, { ReactElement, useEffect } from "react";
import { useSelector } from "react-redux";

import store, { StoreStateType, initializeStore } from "store";
import Box from "components/Box";
import TextField from "components/TextField";
import Calendar from "components/Calendar";
import Button from "components/Button";
import { ReactComponent as IconPin } from "assets/pin.svg";
import { ReactComponent as IconCursor } from "assets/cursor.svg";
import { ReactComponent as IconGuests } from "assets/guests.svg";

import { FieldSet, GuestsContainer } from "App/styles";

interface AppFormProps {
  openGuests: () => void;
}

const AppForm = ({ openGuests }: AppFormProps): ReactElement => {
  const { rooms } = useSelector((state: StoreStateType) => state);
  const urlParams = new URLSearchParams(window.location.search);

  useEffect(() => {
    store.dispatch(initializeStore(urlParams.get("rooms")));
  }, []);

  return (
    <Box>
      <form>
        <FieldSet>
          <TextField
            placeholder="Type city, place, or hotel name"
            leftIcon={<IconPin />}
            rightIcon={<IconCursor />}
            onChange={() => {}}
          />
        </FieldSet>
        <FieldSet>
          <Calendar />
          <GuestsContainer>
            <TextField
              leftIcon={<IconGuests />}
              type="number"
              min={1}
              max={9}
              readonly
              value={String(rooms.length)}
              onClick={openGuests}
              onChange={() => {}}
            />
          </GuestsContainer>
        </FieldSet>
        <Button onClick={() => {}}>Search</Button>
      </form>
    </Box>
  );
};

export default AppForm;
