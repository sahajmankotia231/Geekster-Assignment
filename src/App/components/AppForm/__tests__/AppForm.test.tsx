import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

import store from "store";

import AppForm from "..";

describe("<App />", () => {
  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <AppForm openGuests={() => {}} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
