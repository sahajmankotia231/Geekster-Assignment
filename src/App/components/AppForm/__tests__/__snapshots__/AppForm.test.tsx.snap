// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render properly 1`] = `
<div
  class="sc-bdvvaa dfsOak"
>
  <form>
    <fieldset
      class="sc-crHlIS gYBSAF"
    >
      <div
        class="sc-gsDJrp iOGeOr"
      >
        <i
          class="sc-dkPtyc dBwcwA"
        >
          <svg>
            pin.svg
          </svg>
        </i>
        <input
          class="sc-eCImvq fOIGbL"
          placeholder="Type city, place, or hotel name"
          type="text"
          value=""
        />
        <i
          class="sc-hKwCoD cXzwwS"
        >
          <svg>
            cursor.svg
          </svg>
        </i>
      </div>
    </fieldset>
    <fieldset
      class="sc-crHlIS gYBSAF"
    >
      <div
        class="sc-jRQAMF ilBMk"
      >
        <svg>
          calendar.svg
        </svg>
        <strong>
          Check-in
        </strong>
        <span
          class="sc-gKckTs JtJqW"
        />
        <strong>
          Check-out
        </strong>
      </div>
      <div
        class="sc-egiSv izdmIH"
      >
        <div
          class="sc-gsDJrp iOGeOr"
        >
          <i
            class="sc-dkPtyc dBwcwA"
          >
            <svg>
              guests.svg
            </svg>
          </i>
          <input
            class="sc-eCImvq izJtxO"
            max="9"
            min="1"
            type="number"
            value="1"
          />
        </div>
      </div>
    </fieldset>
    <button
      class="sc-iCfLBT fgyNbE"
      data-testid="button"
    >
      Search
    </button>
  </form>
</div>
`;
